### PR TITLE
Move to non-segmented dictionary

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
ImmutableSegmentedDictionary can exhibit n^2 memory performance on linear mutation (e.g. each mutation is O(n)).  Moving to ImmutableDictionary as that has n-log(n) performance.  This addresses the original allocations of:

image

that was fixed in https://github.com/dotnet/roslyn/pull/67655.  However, it hopefully does not exhibit the poor scaling that Todd observed in other scenarios.  He is testing that now.

### CODE REVIEW

**Comment by sharwell** [comment]
❔ In general is this map read many more times than it's mutated? If so, we may see an execution time hit due to poor lookup performance. We never iterate over this map, so iteration performance isn't going to be a problem.

**Comment by CyrusNajmabadi** [comment]
> ❔ In general is this map read many more times than it's mutated?

No.  :)

> If so, we may see an execution time hit due

It is called when a binder is freshly made at the level of a namespace that contains usings.  The binders are made once per semantic model.   Haven't seen any perf hit on this, and this retains the memory win my original PR had (and todd is validating it addressed the n^2 memory hit he was seeing in his scenario) :)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs [hunk 1]
```diff
@@ -33,14 +36,12 @@ internal sealed partial class SourceNamespaceSymbol : NamespaceSymbol
         /// <summary>
         /// Should only be read using <see cref="GetAliasesAndUsings(SingleNamespaceDeclaration)"/>.
         /// </summary>
-        private ImmutableSegmentedDictionary<SingleNamespaceDeclaration, AliasesAndUsings> _aliasesAndUsings_doNotAccessDirectly =
-            ImmutableSegmentedDictionary<SingleNamespaceDeclaration, AliasesAndUsings>.Empty.WithComparer(ReferenceEqualityComparer.Instance);
+        private ImmutableDictionary<SingleNamespaceDeclaration, AliasesAndUsings> _aliasesAndUsings_doNotAccessDirectly = s_emptyMap;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs:39]: this was an unfortunate surprise.  I had mentally internalized these two types were effectively interchangeable wrt perf characteristics.

**Comment by CyrusNajmabadi** [comment]
Tagging @333fred @RikkiGibson

**Review by RikkiGibson — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@333fred ptal :)

**Comment by CyrusNajmabadi** [comment]
@jaredpar ptal.

**Pull request merged**